// redis-smq/dist/src/common/redis-client/redis-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ELuaScriptName = void 0;
const redis_smq_common_1 = require("redis-smq-common");
const fs = require("fs");
var ELuaScriptName;
(function (ELuaScriptName) {
    ELuaScriptName["ENQUEUE_SCHEDULED_MESSAGE"] = "ENQUEUE_SCHEDULED_MESSAGE";
    ELuaScriptName["PUBLISH_MESSAGE"] = "PUBLISH_MESSAGE";
    ELuaScriptName["REQUEUE_MESSAGE"] = "REQUEUE_MESSAGE";
    ELuaScriptName["SCHEDULE_MESSAGE"] = "SCHEDULE_MESSAGE";
    ELuaScriptName["HAS_QUEUE_RATE_EXCEEDED"] = "HAS_QUEUE_RATE_EXCEEDED";
    ELuaScriptName["CREATE_QUEUE"] = "CREATE_QUEUE";
    ELuaScriptName["INIT_CONSUMER_QUEUE"] = "INIT_CONSUMER_QUEUE";
})(ELuaScriptName = exports.ELuaScriptName || (exports.ELuaScriptName = {}));
redis_smq_common_1.RedisClient.addScript(ELuaScriptName.ENQUEUE_SCHEDULED_MESSAGE, fs.readFileSync(`${__dirname}/lua/enqueue-scheduled-message.lua`).toString());
redis_smq_common_1.RedisClient.addScript(ELuaScriptName.PUBLISH_MESSAGE, fs.readFileSync(`${__dirname}/lua/publish-message.lua`).toString());
redis_smq_common_1.RedisClient.addScript(ELuaScriptName.REQUEUE_MESSAGE, fs.readFileSync(`${__dirname}/lua/requeue-message.lua`).toString());
redis_smq_common_1.RedisClient.addScript(ELuaScriptName.SCHEDULE_MESSAGE, fs.readFileSync(`${__dirname}/lua/schedule-message.lua`).toString());
redis_smq_common_1.RedisClient.addScript(ELuaScriptName.HAS_QUEUE_RATE_EXCEEDED, fs.readFileSync(`${__dirname}/lua/has-queue-rate-exceeded.lua`).toString());
redis_smq_common_1.RedisClient.addScript(ELuaScriptName.CREATE_QUEUE, fs.readFileSync(`${__dirname}/lua/create-queue.lua`).toString());
redis_smq_common_1.RedisClient.addScript(ELuaScriptName.INIT_CONSUMER_QUEUE, fs.readFileSync(`${__dirname}/lua/init-consumer-queue.lua`).toString());
