// redis-smq-common/dist/src/worker/worker-runner/worker-pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkerPool = void 0;
const async_1 = require("../../async/async");
class WorkerPool {
    constructor() {
        this.pool = [];
        this.index = 0;
        this.getCurrentPoolItem = () => {
            if (this.pool.length) {
                const worker = this.pool[this.index];
                this.index += 1;
                if (this.index >= this.pool.length) {
                    this.index = 0;
                }
                return worker;
            }
            return null;
        };
        this.work = (cb) => {
            const worker = this.getCurrentPoolItem();
            if (worker)
                worker.work(cb);
            else
                cb();
        };
        this.add = (worker) => {
            this.pool.push(worker);
            return this.pool.length;
        };
        this.clear = (cb) => {
            async_1.async.each(this.pool, (worker, _, done) => {
                worker.quit(done);
            }, () => {
                this.pool = [];
                this.index = 0;
                cb();
            });
        };
    }
}
exports.WorkerPool = WorkerPool;
