// redis-smq/dist/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMessageUnacknowledgedCause = exports.EMessageDeadLetterCause = void 0;
var EMessageDeadLetterCause;
(function (EMessageDeadLetterCause) {
    EMessageDeadLetterCause["TTL_EXPIRED"] = "ttl_expired";
    EMessageDeadLetterCause["RETRY_THRESHOLD_EXCEEDED"] = "retry_threshold_exceeded";
    EMessageDeadLetterCause["PERIODIC_MESSAGE"] = "periodic_message";
})(EMessageDeadLetterCause = exports.EMessageDeadLetterCause || (exports.EMessageDeadLetterCause = {}));
var EMessageUnacknowledgedCause;
(function (EMessageUnacknowledgedCause) {
    EMessageUnacknowledgedCause["TIMEOUT"] = "timeout";
    EMessageUnacknowledgedCause["CONSUME_ERROR"] = "consumer_error";
    EMessageUnacknowledgedCause["UNACKNOWLEDGED"] = "unacknowledged";
    EMessageUnacknowledgedCause["OFFLINE_CONSUMER"] = "offline_consumer";
    EMessageUnacknowledgedCause["TTL_EXPIRED"] = "ttl_expired";
})(EMessageUnacknowledgedCause = exports.EMessageUnacknowledgedCause || (exports.EMessageUnacknowledgedCause = {}));
